gitter.py

```diff
@@ -297,9 +300,14 @@ def background():
                 else:
                     log.debug('Received keep-alive on %s', room.name)
 
-        t = threading.Thread(target=background)
-        t.daemon = True
-        t.start()
+        if room._uri not in self._joined_rooms:
+            t = threading.Thread(target=background)
+            t.daemon = True
+            t.start()
+            self._joined_rooms.append(room._uri)
+        else:
+            with self._joined_rooms_lock:
+                log.info("Already joined the room.")
```

Comment:
The thread lock (`with self._joined_rooms_lock:`) is in the wrong place here. It needs to guard access and modification of `self._joined_rooms` (which happens in `if room._uri not in self._joined_rooms` and `self._joined_rooms.append(room._uri)`.